haskell/test/SMT.hs

```diff
@@ -0,0 +1,176 @@
+import Hedgehog
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.Hedgehog (testProperty)
+
+import Lorentz hiding (now, (>>))
+import Michelson.Typed.Convert (convertContract, untypeValue)
+import Morley.Nettest
+import qualified Morley.Micheline as MM
+import Morley.Nettest.Pure
+
+import Ligo.BaseDAO.Contract
+import Ligo.BaseDAO.Types
+import Test.Ligo.BaseDAO.Common.StorageHelper as StorageHelper
+import Test.Ligo.BaseDAO.Common
+import SMT.BaseDAO.Management
+import SMT.BaseDAO.Proposal
+import SMT.BaseDAO.Token
+import SMT.Gen
+import SMT.Types
+
+-- originate onece and call multiple time
+test_SMT :: TestTree
+test_SMT = testGroup "SMT"
+  [ testProperty "SMT" $
+      withTests 20 $ property $ do
+        PropertyTestInput (inputs, cs) <- forAll genPropertyTestInput
+        nettestTestProp $
+          (uncapsNettestEmulated $ do
+              -- Set initial level
+              advanceLevel (cs & csBlockchainState & bsLevel)
+
+              guardianContract <- originateSimple "guardian" () dummyGuardianContract
+              tokenContract <- originateSimple "TokenContract" [] dummyFA2Contract
+
+              let newCs = addAddresses cs (unTAddress guardianContract) (unTAddress tokenContract)
+
+              -- Originate Dao for nettest
+              let
+                originateData = UntypedOriginateData
+                  { uodName = "BaseDAO"
+                  , uodBalance = toMutez 0
+                  , uodStorage = untypeValue $ toVal $ (newCs & csStorage)
+                  , uodContract = convertContract baseDAOContractLigo
+                  }
+
+              daoUntyped <- originateUntyped originateData
+              let dao = TAddress @Parameter daoUntyped
+
+              callEntrypointAndCompare dao inputs newCs
+
+          )
+          emulatedImplIntegrational
+  ]
+  where
+    -- Reason: guardian address and governance token address are not known until we originate the contract in nettest
+    addAddresses cs guardAddr govAddr = cs
+        { csStorage = (cs & csStorage)
+            { fsStorage = (cs & csStorage & fsStorage)
+                { sGuardian = guardAddr
+                , sGovernanceToken = (cs & csStorage & fsStorage & sGovernanceToken)
+                    { gtAddress = govAddr
+                    }
+                }
+            }
+        }
+
+callEntrypointAndCompare :: TAddress Parameter -> [ContractCall Parameter] -> ContractState -> EmulatedT PureM ()
+callEntrypointAndCompare _ [] _ = pure ()
+callEntrypointAndCompare dao (cc:ccs) cs = do
+  let haskellResult = callBaseDaoHaskell cc cs
+  ligoResult <- callBaseDaoLigo dao cc cs
+
+  -- Note: `fsConfig` can't be compared because using `getStorage` resulted in different
+  -- different lambda.
+  -- Ex: cRejectedProposalSlashValue:
+  -- - from haskell: `Seq (Seq (DROPN (SS (SS SZ))) (PUSH (VNat 0))) Nop`
+  -- - from ligo: ` Seq (AnnPAIR : % %) (Seq (AnnUNPAIR @ @ % %) (Seq (DROPN (SS (SS SZ))) (PUSH (VNat 0))))`
```

Comment:
Initially, they are the same, but after calling an entrypoints in Nettest, and do a `getStorage`, the lambda starts to contain `AnnPair` (semantically, it should still be the same). I think it may have to do with how `getStorage` work.

Reply:
Well this sounds like an issue coming from `morley`, although I'm not entirely sure it should be considered a bug.

The empty anns should not be considered in the `Eq` instance I think, we'll check it out.

In the meanwhile you can keep comparing not comparing the config, which should be ok as it should never be updated IIRC.

Reply:
Yeah, I suspect it is from `morley` as well. I didn't spend much time investigate into this and simply not comparing the `config` for now.

Reply:
Sounds good to me, config should never change anyway :ok_hand: